### PR TITLE
[move-prover] dedicated syntax for loop invariants

### PR DESCRIPTION
**Pull request by mengxu-fb**
With the fine-grained separation of `ConditionKind::Invariant`, we can
finally use the `invariant` keyword to mark loop invariants.

The prior approach is to use `assert` to mark a loop invariant and
subsequently check whether this `assert` appears in a loop header
block. If an `assert` is in the loop header block, it is considered as
a loop invariant. Otherwise, it is considered as a normal assert.

The difference between a loop invariant and a normal assert is that
a loop invariant will be translated into three statements:
- an `assert` in the beginning of the header block
- an `assume` after `havoc`-ing loop induction variables,
- an `assert` in the induction case checking block

While a normal assert will only be translated as an `assert` towards
the end of the loop header block.

It is better to review this PR commit by commit,
- the first commit contains the core logic,
- the subsequent `fixup!` commits are either mechanical changes
  of `assert -> invariant` or auto-updated `.md` files
- the last commit is a test case on invalid declarations of
  `invariant` in the code.

## Motivation

Separate the loop invariants and normal asserts in the code.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

- CI
- new tests added

### CODE REVIEW

**Comment by shazqadeer** [comment]
@mengxu-fb : Can asserts in the loop header block still be used as an invariant?  For example, what happens if the programmer wrote a sequence of invariants followed by a sequence of asserts?

**Comment by mengxu-fb** [comment]
> Can asserts in the loop header block still be used as an invariant?

No, the asserts are just normal asserts, they will not be considered as loop invariants, i.e.,
- they will not be checked on the loop base case
- they will not be assumed after havoc
- they will not be checked on the induction case.

> what happens if the programmer wrote a sequence of invariants followed by a sequence of asserts?
```
invariant I1;
...
invariant In;

assert P;
```
Will be translated to
```
assert I1 ... In;
havoc;
assume I1 ... In;

assert P;
```

So, this is essentially saying, whether `I1 & ... & In & (other conditions) ==> P`. This can be viewed as a debugging tool to see whether we have put strong enough invariants to prove some property with the invariants.

File: language/move-prover/bytecode/src/loop_analysis.rs [hunk 1]
```diff
@@ -251,31 +251,40 @@ impl LoopAnalysisProcessor {
...
+        // we have unrolled the loop into a DAG, and there will be no loop invariants left
+        builder.data.loop_invariants.clear();
         builder.data
     }
 
     /// Collect invariants in the given loop header block
     ///
-    /// Loop invariants are defined as the longest sequence of consecutive 'assert'
-    /// statements in the loop header block, immediately after the Label statement.
+    /// Loop invariants are defined as the longest sequence of consecutive 'assert' statements in
```
Comment (shazqadeer) [inline language/move-prover/bytecode/src/loop_analysis.rs:261]: Shouldn't this comment say: ...longest sequence of consecutive `invariant` statements?

Comment (mengxu-fb) [reply language/move-prover/bytecode/src/loop_analysis.rs:261]: My bad, yes, the longest sequence of consecutive `invariant` statements.

File: language/move-prover/tests/sources/functional/loop_invariant_invalid.move [hunk 2]
```diff
@@ -0,0 +1,29 @@
+module 0x42::InvalidLoopInvariants {
```
Comment (shazqadeer) [inline language/move-prover/tests/sources/functional/loop_invariant_invalid.move:1]: Please add couple more test cases to show that it is not allowed to have assume/asserts before the invariants? Also, I was wondering if it is possible to have only assume/assert statements in the loop header without any invariants?

File: language/move-prover/bytecode/src/loop_analysis.rs (hunk 1 above)
Comment (mengxu-fb) [reply language/move-prover/bytecode/src/loop_analysis.rs:261]: Hmm, this is a bit tricky. Given that we don't have an `invariant` type in our `PropKind` (all we have is `assert`, `assume`, and `modifies`), the `invariant` type is simulated by the combination of
- the `PropKind::Assert` and
- its `AttrId` exists in `FunctionData.loop_invariants`, which is a set of `AttrId`.

I'll update the comment to be more explicit on this.

File: language/move-prover/tests/sources/functional/loop_invariant_invalid.move (hunk 2 above)
Comment (mengxu-fb) [reply language/move-prover/tests/sources/functional/loop_invariant_invalid.move:1]: > if it is possible to have only assume/assert statements in the loop header without any invariants?

Yes, this is allowed.

Comment (shazqadeer) [reply language/move-prover/tests/sources/functional/loop_invariant_invalid.move:1]: In that case, perhaps add test case/s to illustrate that assume/assert can occur by themselves in the loop header block.

**Review by shazqadeer — DISMISSED** [review]

**Review by shazqadeer — APPROVED** [review]

**Pull request closed without merging**